src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -217,53 +222,68 @@ public FormPipeReader(PipeReader pipeReader, Encoding encoding)
             bool isFinalBlock)
         {
             var sequenceReader = new SequenceReader<byte>(buffer);
+            ReadOnlySequence<byte> segment;
+
             var consumed = sequenceReader.Position;
             var consumedBytes = default(long);
             var equalsDelimiter = GetEqualsForEncoding();
             var andDelimiter = GetAndForEncoding();
 
             while (!sequenceReader.End)
             {
-                // TODO seems there is a bug with TryReadTo (advancePastDelimiter: true). It isn't advancing past the delimiter on second read.
-                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> key, equalsDelimiter, advancePastDelimiter: false) ||
-                    !sequenceReader.IsNext(equalsDelimiter, true))
+                if (!sequenceReader.TryReadTo(out segment, andDelimiter))
                 {
-                    if ((sequenceReader.Consumed - consumedBytes) > KeyLengthLimit)
+                    if (!isFinalBlock)
                     {
-                        ThrowKeyTooLargeException();
+                        // Don't buffer indefinately
+                        if ((sequenceReader.Consumed - consumedBytes) > KeyLengthLimit + ValueLengthLimit)
+                        {
+                            ThrowKeyOrValueTooLargeException();
+                        }
+                        break;
                     }
 
-                    break;
+                    // This must be the final key=value pair
+                    segment = buffer.Slice(sequenceReader.Position);
+                    sequenceReader.Advance(segment.Length);
                 }
 
-                if (key.Length > KeyLengthLimit)
+                if (segment.IsSingleSegment)
                 {
-                    ThrowKeyTooLargeException();
+                    ParseFormValuesFast(segment.First.Span, ref accumulator, isFinalBlock: true, out var segmentConsumed);
```

Comment:
Very nice change, I was looking at this earlier (if we should swap to the fast path). You can use FirstSpan instead of First.Span (it’s new in 3.0)